Java/core/src/main/java/com/amazon/randomcutforest/store/StoreCommon.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.store;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.CommonUtils.checkState;
+
+import java.util.BitSet;
+
+/**
+ * This class defines common functionality for Store classes, including
+ * maintaining the stack of free pointers.
+ */
+public class StoreCommon {
```

Comment:
1/ Should this class be an abstract class? 

2/ This and the Store class looks very similar. Should we abstract out the common parts into a base class? This abstraction can also be extended by the SmallStoreCommon.. 

3/ Or Should we make a concrete class called IndexStore? This could just maintain the indexes, the same logic can be used by other Storage classes which want to maintain the indexes? (Still need to think about this)

Reply:
This PR contains two completely independent implementations of the store concept:  one based on generics, and one using primitive arrays. So eventually, we won't have both Store<T> and StoreCommon.

The StoreCommon class can be used as the base class for a column-oriented store class that uses ints as indexes. Currently I don't think there are any methods that we would want to make abstract. The best candidate would be an abstract `add` method, but the problem with that is that the add methods in different subclasses will have different method signatures depending on the type of data being stored. For example, compare `NodeStore::add` to `LeafStore::add`.

StoreCommon contains just the logic to keep track of what indexes are in use or not. It's similar to the other comment about making ReferenceCounter a separate class: instead of subclassing StoreCommon, another approach could be to create a storage class that keeps an instance of StoreCommon for tracking indexes. And similar to that case, I don't have a strong opoinion about which approach is better.

Reply:
if this class is only keeping track of what indexes are in use, should we rename it?

Reply:
In the latest update, I changed the base classes to IndexManager and SmallIndexManager. I opted to keep have the store classes inherit from them (as opposed to having the store classes create an instance of IndexManager), because it makes sense for those classes to inherit the size() and getCapacity() methods.